src/bynder-js-sdk.js

```diff
@@ -597,4 +599,347 @@ export default class Bynder {
+            saveURL,
+            'POST',
+            this.consumerToken,
+            this.accessToken,
+            data
+        );
+        return request.send();
+    }
+
+    /**
+     * Uploads arbirtrarily sized buffer or stream file to provided S3 endpoint in chunks and registers each chunk to Bynder.
+     * Resolves the passed init result and final chunk number.
+     * @param {Object} file ={} - An object containing the id of the desired collection.
+     * @param {String} file.filename - The file name of the file to be saved
+     * @param {Buffer|Readable} file.body - The file to be uploaded. Can be either buffer or a read stream.
+     * @param {Number} file.length - The length of the file to be uploaded
+     * @param {string} endpoint - S3 endpoint url
+     * @param {Object} init - Result from init upload
+     * @return {Promise}
+     */
+    uploadFileInChunks(file, endpoint, init) {
+        const { body } = file;
+        const isBuffer = Buffer.isBuffer(body);
+        const length = isBuffer ? body.length : file.length;
+        const CHUNK_SIZE = 1024 * 1024 * 5;
+        const chunks = Math.ceil(length / CHUNK_SIZE);
+
+        const registerChunk = this.registerChunk.bind(this);
+
+        const uploadChunkToS3 = (chunkData, chunkNumber) => {
+            const form = new FormData();
+            const uploadPath = init.multipart_params.key;
+            const params = Object.assign(init.multipart_params, {
+                name: `${basename(uploadPath)}/p${chunkNumber}`,
+                chunk: chunkNumber,
+                chunks,
+                Filename: `${uploadPath}/p${chunkNumber}`,
+                key: `${uploadPath}/p${chunkNumber}`
+            });
+            Object.keys(params).forEach((key) => {
+                form.append(key, params[key]);
+            });
+            form.append('file', chunkData);
+            const headers = Object.assign(form.getHeaders(), {
+                'content-length': form.getLengthSync()
+            });
+            return axios.post(endpoint, form, { headers });
+        };
+
+        // sequentially upload chunks to AWS, then register them
+        return new Promise((resolve, reject) => {
+            let chunkNumber = 0;
+            (function nextChunk() {
+                if (chunkNumber >= chunks) {
+                    // we are finished, pass init and chunk number to be finalised
+                    return resolve({ init, chunkNumber });
+                }
+
+                // upload next chunk
+                let chunkData;
+                if (isBuffer) {
+                    // handle buffer data
+                    const start = chunkNumber * CHUNK_SIZE;
+                    const end = Math.min(start + CHUNK_SIZE, length);
+                    chunkData = body.slice(start, end);
+                } else {
+                    // handle stream data
+                    chunkData = body.read(CHUNK_SIZE);
+                    if (chunkData === null) {
```

Comment:
I searched a bit about this and all the examples I could find are implemented like this:
```
while (null !== (chunkData = body.read(CHUNK_SIZE))) {
    break;
}
```
I'm having an hard time deciding which one is more efficient or elegant, what do you think?

Reply:
A while loop will completely block the event loop. setTimeout is much better here, as it is asynchronous and doesn't block.